openwisp_controller/connection/connectors/snmp.py

```diff
@@ -0,0 +1,51 @@
+import logging
+
+from django.utils.functional import cached_property
+from jsonschema import validate
+from jsonschema.exceptions import ValidationError as SchemaError
+
+logger = logging.getLogger(__name__)
+
+
+class Snmp(object):
+    schema = {
+        '$schema': 'http://json-schema.org/draft-04/schema#',
+        'type': 'object',
+        'title': 'Credentials type',
+        'oneOf': [
+            {
+                'title': 'SNMP',
+                'required': ['community', 'agent'],
+                'additionalProperties': False,
+                'properties': {
+                    'community': {'type': 'string', 'default': 'public'},
+                    'agent': {'type': 'string'},
+                    'port': {
+                        'type': 'integer',
+                        'default': 161,
+                        'minimum': 1,
+                        'maximum': 65535,
+                    },
+                },
+            },
+        ],
+    }
+
+    def __init__(self, params, addresses):
+        self._params = params
+        self.addresses = addresses
+
+    @classmethod
+    def validate(cls, params):
+        validate(params, cls.schema)
+        cls.custom_validation(params)
+
+    @classmethod
+    def custom_validation(cls, params):
+        if 'community' not in params or 'agent' not in params:
+            raise SchemaError('Missing community or agent')
+
+    @cached_property
+    def params(self):
+        params = self._params.copy()
+        return params
```

Comment:
I don't think it will make any difference if we just use `self.params` here. We don't even need to store `_params. What do you think?